cpp/src/arrow/array.cc

```diff
@@ -982,4 +982,284 @@ std::vector<ArrayVector> RechunkArraysConsistently(
 
 }  // namespace internal
 
+struct ConcatenateImpl {
```

Comment:
This implementation needs a more appropriate home than array.cc. We could create a subdirectory of miscellaneous array algorithms. I don't have a strong opinion, @xhochy @pitrou @fsaintjacques @emkornfield thoughts?

Reply:
under util/ might make sense for now.  We can see how many random algorithms get put here vs just used in kernels (my second choice would be under compute I suppose).

Reply:
Algorithms could be grouped under `src/arrow/array/algorithm_<category>.{h,cc}` with a single convenience header/source `src/arrow/algorithm.{h,cc}` similar to builders

Reply:
If this is all private, we should avoid convenience headers that balloon compile times.